### PR TITLE
Move files from temporary directory changer

### PR DESCRIPTION
**Pull request by mgjarrett**
## What is the change?

Use `shutil.move` / `mv` instead of `shutil.copy` / `cp` to move files from temporary directory back to the original working directory on `__exit__`.

## Why is the change being made?

Copying files from one location to another keeps the original file intact, but the original files in a temporary directory do not need to be kept around after the `__exit__` from the `TemporaryDirectoryChanger`. These files are immediately deleted after the copy operation is completed.

Using `shutil.move()` is significantly faster than `shutil.copy()` when the source and destination locations are on the same drive, since the data doesn't have to physically be copied to a new place in memory.

---

## Checklist

- [x] This PR has only [one purpose or idea](https://terrapower.github.io/armi/developer/tooling.html#one-idea-one-pr).
- [x] [Tests](https://terrapower.github.io/armi/developer/tooling.html#test-it) have been added/updated to verify any new/changed code.

- [x] The code style follows [good practices](https://terrapower.github.io/armi/developer/standards_and_practices.html).
- [x] The commit message(s) follow [good practices](https://terrapower.github.io/armi/developer/tooling.html).

- [x] The [release notes](https://terrapower.github.io/armi/developer/tooling.html#add-release-notes) have been updated if necessary.
- [x] The [documentation](https://terrapower.github.io/armi/developer/tooling.html#document-it) is still up-to-date in the `doc` folder.
- [x] The dependencies are still up-to-date in `pyproject.toml`.

### CODE REVIEW

File: armi/utils/__init__.py [hunk 1]
```diff
@@ -837,6 +837,49 @@ def safeCopy(src: str, dst: str) -> None:
     runLog.extra("Copied {} -> {}".format(src, dst))
 
 
+def safeMove(src: str, dst: str) -> None:
```
Comment (john-science) [inline armi/utils/__init__.py:840]: Can we get a new unit test for this new function?

If it helps I see the test go something like this:

* Use the `TemporaryDirectoryChanger` to create a temp dir for your test
* create two dirs, one called `src` and one called `dst`
* add a blank fie or two inside `src` (or just add 1 character of text inside them)
* test 1: use this function to move the files from `src/` into `dst`
* test 2: use this function to move the entire `src/` dir into `dst`

Something like that, anyway.

**Review by opotowsky — COMMENTED** [review]
Looks like the move doesn't work with windows/mac. 

But I'd like to make a higher level comment. The secret overwriting of shutil.copy is something I want to remove from ARMI (i.e., I want to go change all the instances of `shutil.copy` to explicitly use `safeCopy` instead, then remove the overwrite. See #1690.

Maybe for this update it makes sense to keep the same scheme and we just tack this onto that tech debt payoff since it won't really be causing any extra work.

But I'd also be happy if the `shutil.move` calls were just `safeMove` instead. :-)

**Comment by mgjarrett** [comment]
I didn't touch `safeCopy` at all, but I added some `assert` statements to verify that the `safeCopy` call is actually copying the files. The unit test for `safeCopy` passes locally for me, but here it is now failing on those assertions, which is strange.

It's also weird that the existing unit test of `safeCopy` doesn't actually assert that the files are copied successfully. Makes me wonder if there is a known issue here.

File: armi/utils/__init__.py (hunk 1 above)
Comment (mgjarrett) [reply armi/utils/__init__.py:840]: I added some unit tests

File: armi/utils/__init__.py [hunk 2]
```diff
@@ -837,6 +837,47 @@ def safeCopy(src: str, dst: str) -> None:
...
+    if "win" in sys.platform:
+        shutil.move(src, dst)
+    elif "linux" in sys.platform:
+        cmd = f'mv "{src}" "{dst}"'
+        os.system(cmd)
+    else:
+        raise OSError(
+            "Cannot perform ``safeMove`` on files because ARMI only supports "
+            + "Linux and Windows."
+        )
```
Comment (opotowsky) [inline armi/utils/__init__.py:857]: We run unit tests on mac OS now. So is this OSError (here and in other places) accurate? @john-science

Comment (opotowsky) [reply armi/utils/__init__.py:857]: (keep in mind Mike is just following the "template" created in `safeCopy`)

**Review by opotowsky — APPROVED** [review]
This looks great!

File: armi/utils/__init__.py (hunk 2 above)
Comment (john-science) [reply armi/utils/__init__.py:857]: @mgjarrett ARMI CI runs on Mac, Linux, and Windows. Yet I don't see this error being raised when the unit tests run on MacOS.  (Your CI passes in this PR.)

What's going on?

Comment (john-science) [reply armi/utils/__init__.py:857]: Also, I note, every single other feature in ARMI supports the MacOS. I'm not sure why this would be different.

Comment (opotowsky) [reply armi/utils/__init__.py:857]: @john-science before macos was added, we added this exact if/elif/else configuration to several places in ARMI. You'll note it is in `safeCopy` just above this in this file.

I'm not sure why the test passes on macos, but note that `test_safeCopy` also passes and it's the same structure.

**Comment by opotowsky** [comment]
> > The secret overwriting of shutil.copy is something I want to remove from ARMI
> 
> I note we already have a ticket for this: #1690
> 
> But I think Michael's PR doesn't touch that, because he didn't alter `safeCopy()`, but added a new `saveMove()`. Am I wrong?

Yep! the first iteration in this PR had a similar setup, (where shutil.move was overwritten with `safeMove`) but that was undone in favor of not overwriting `shutil.move`.

File: armi/utils/__init__.py [hunk 3]
```diff
@@ -837,6 +837,47 @@ def safeCopy(src: str, dst: str) -> None:
...
+    srcSize = os.path.getsize(src)
+    if "win" in sys.platform:
+        shutil.move(src, dst)
+    elif "linux" in sys.platform:
+        cmd = f'mv "{src}" "{dst}"'
+        os.system(cmd)
+    else:
+        raise OSError(
+            "Cannot perform ``safeMove`` on files because ARMI only supports "
+            + "Linux and Windows."
```
Comment (john-science) [inline armi/utils/__init__.py:856]:
```suggestion
            + "Linux, MacOS, and Windows."
```

File: armi/utils/__init__.py [hunk 4]
```diff
@@ -837,6 +837,47 @@ def safeCopy(src: str, dst: str) -> None:
...
+def safeMove(src: str, dst: str) -> None:
+    """Check that a file has been successfully moved before continuing."""
+    # Convert files to OS-independence
+    src = os.path.abspath(src)
+    dst = os.path.abspath(dst)
+    if os.path.isdir(dst):
+        dst = os.path.join(dst, os.path.basename(src))
+    srcSize = os.path.getsize(src)
+    if "win" in sys.platform:
+        shutil.move(src, dst)
```
Comment (john-science) [inline armi/utils/__init__.py:849]:
```suggestion
    if os.path.isdir(dst):
        dst = os.path.join(dst, os.path.basename(src))

    srcSize = os.path.getsize(src)
    if "win" in sys.platform:
        shutil.move(src, dst)
```

File: armi/utils/__init__.py [hunk 5]
```diff
@@ -837,6 +837,48 @@ def safeCopy(src: str, dst: str) -> None:
...
+    """Check that a file has been successfully moved before continuing."""
+    # Convert files to OS-independence
+    src = os.path.abspath(src)
+    dst = os.path.abspath(dst)
+    if os.path.isdir(dst):
+        dst = os.path.join(dst, os.path.basename(src))
+
+    srcSize = os.path.getsize(src)
+    if "win" in sys.platform:
+        shutil.move(src, dst)
```
Comment (john-science) [inline armi/utils/__init__.py:850]:
```suggestion
    if "win" in sys.platform:
        # this covers Windows ("win32") and MacOS ("darwin")
        shutil.move(src, dst)
```

File: armi/utils/__init__.py [hunk 6]
```diff
@@ -837,6 +837,48 @@ def safeCopy(src: str, dst: str) -> None:
...
+    elif "linux" in sys.platform:
+        cmd = f'mv "{src}" "{dst}"'
+        os.system(cmd)
+    else:
+        raise OSError(
+            "Cannot perform ``safeMove`` on files because ARMI only supports "
+            + "Linux, MacOS, and Windows."
+        )
+    waitTime = 0.01  # 10 ms
```
Comment (john-science) [inline armi/utils/__init__.py:859]:
```suggestion
        )

    waitTime = 0.01  # 10 ms
```

File: armi/utils/__init__.py [hunk 7]
```diff
@@ -837,6 +837,48 @@ def safeCopy(src: str, dst: str) -> None:
...
+        time.sleep(waitTime)
+        totalWaitTime += waitTime
+        if totalWaitTime > maxWaitTime:
+            runLog.warning(
+                f"File move from {dst} to {src} has failed due to exceeding "
+                + f"a maximum wait time of {maxWaitTime/60} minutes."
+            )
+            break
+
+    runLog.extra("Moved {} -> {}".format(src, dst))
```
Comment (john-science) [inline armi/utils/__init__.py:878]: Okay, this print-out that says the move was successful will happen even if the move times out.

Let me fix that.

File: armi/utils/__init__.py [hunk 8]
```diff
@@ -837,6 +837,48 @@ def safeCopy(src: str, dst: str) -> None:
...
+        except FileNotFoundError:
+            pass
+        time.sleep(waitTime)
+        totalWaitTime += waitTime
+        if totalWaitTime > maxWaitTime:
+            runLog.warning(
+                f"File move from {dst} to {src} has failed due to exceeding "
+                + f"a maximum wait time of {maxWaitTime/60} minutes."
+            )
+            break
```
Comment (john-science) [inline armi/utils/__init__.py:876]:
```suggestion
            return
```

**Review by john-science — APPROVED** [review]

**Pull request merged**